src/printer.js

```diff
@@ -325,76 +325,95 @@ function printMemberChain(path, options, print) {
-  // propertylookup
-  let hasSeenCallExpression = false;
-  for (; i < printedNodes.length; ++i) {
-    if (hasSeenCallExpression && isMemberish(printedNodes[i].node)) {
-      // [0] should be appended at the end of the group instead of the
-      // beginning of the next one
+
+      // for the first node, we have some special cases to dictate whether to
+      // group following nodes or put it on its own line.
+      //
+      // group static lookups
+      // SomeClass::staticFunction()
+      //   ->someOtherThing;
+      //
+      // group member lookups, if there are 2 in a row
+      // $this->something->someOtherThing()
+      //   ->someOtherThing;
       if (
-        printedNodes[i].node.kind === "offsetlookup" &&
-        printedNodes[i].node.offset &&
-        printedNodes[i].node.offset.kind === "number"
+        index === 0 &&
+        printedNodes[index + 1] &&
+        printedNodes[index + 1].node.kind !== "call" &&
+        printedNodes[index + 1].node.kind !== "staticlookup" &&
+        !(
+          isMemberish(printedNodes[index + 1].node) &&
+          printedNodes[index + 2] &&
+          isMemberish(printedNodes[index + 2].node)
+        )
       ) {
-        currentGroup.push(printedNodes[i]);
-        continue;
+        shouldStartNewGroup = true;
       }
 
-      groups.push(currentGroup);
-      currentGroup = [];
-      hasSeenCallExpression = false;
-    }
-
-    if (printedNodes[i].node.kind === "call") {
-      hasSeenCallExpression = true;
-    }
-    currentGroup.push(printedNodes[i]);
-
-    if (
-      printedNodes[i].node.comments &&
-      comments.hasTrailingComment(printedNodes[i].node)
-    ) {
-      groups.push(currentGroup);
-      currentGroup = [];
-      hasSeenCallExpression = false;
+      // if this is the first node, create the first group
+      if (index === 0) {
+        return {
+          currentGroupIndex: 0,
+          shouldStartNewGroup,
+          delayedNewGroup,
+          groups: [[currentNode]]
+        };
+      }
+      // the previous iteration dictates if we should be starting a new group
+      // or appending to the current one
+      if (acc.shouldStartNewGroup) {
+        return {
+          currentGroupIndex: acc.currentGroupIndex + 1,
+          shouldStartNewGroup,
+          delayedNewGroup,
+          groups: [...acc.groups, [currentNode]]
+        };
+      }
+      return {
+        currentGroupIndex: acc.currentGroupIndex,
+        shouldStartNewGroup,
+        delayedNewGroup,
+        groups: [
+          ...acc.groups.slice(0, acc.currentGroupIndex),
+          [...acc.groups[acc.currentGroupIndex], currentNode],
+          ...acc.groups.slice(acc.currentGroupIndex + 1)
+        ]
+      };
+    },
+    {
+      currentGroupIndex: 0,
+      shouldStartNewGroup: true,
+      delayedNewGroup: false,
+      groups: []
     }
-  }
-  if (currentGroup.length > 0) {
-    groups.push(currentGroup);
-  }
+  );
```

Comment:
I don't think it is good solution:
1. It breaks compatibility with the original prettier algo
2. A lot of bug and regressions, try to run this on laravel-framework
3. It can be solve in more easy way

Reply:
Ok I thought this simplified things and made it more understandable (i couldn't figure out how the existing one worked), but fair points. Do you have examples for 2? We can add more tests. Also what for 3. what are you thinking?

Reply:
@mgrip i will do PR with fix in tomorrow with tests and example how we solve this in easy way :+1:

Reply:
sorry guys for delay, a lot of work :disappointed: , send the PR tomorrow :star: